File: frontend/src/types/directive.ts
export type DirectiveStatus = 'draft' | 'active' | 'completed' | 'cancelled'
export type DirectivePriority = 'normal' | 'urgent' | 'very_urgent'

export interface UserMin {
  id: number
  username: string
  full_name: string | null
}

export interface StaffMin {
  id: number
  full_name: string
  position: string | null
  employee_code: string | null
}

export interface DeptMin {
  id: number
  name: string
  short_name: string | null
}

export interface DocMin {
  id: number
  doc_number: string | null
  title: string
  doc_type: string
}

export interface TaskMin {
  id: number
  title: string
  status: string
  priority: string
  deadline: string | null
  assignee: UserMin | null
}

export interface DirectiveRead {
  id: number
  title: string
  content: string | null
  issuer_id: number
  status: DirectiveStatus
  priority: DirectivePriority
  issued_date: string | null
  deadline: string | null
  progress: number
  doc_id: number | null
  assignee_staff_id: number | null
  responsible_department_id: number | null
  created_by: number
  issuer: UserMin
  creator: UserMin
  document: DocMin | null
  assignee_staff: StaffMin | null
  responsible_department: DeptMin | null
  created_at: string
  updated_at: string | null
}

export interface DirectiveUnitRead {
  id: number
  unit_name: string
  role: string
  department_id: number | null
  user_id: number | null
  progress: number
  note: string | null
  department: DeptMin | null
  user: UserMin | null
  created_at: string
  updated_at: string | null
}

export interface DirectiveTaskRead {
  id: number
  task: TaskMin
  created_at: string
}

export interface DirectiveCommentRead {
  id: number
  content: string
  user: UserMin
  created_at: string
}

export interface DirectiveHistoryRead {
  id: number
  action: string
  old_status: string | null
  new_status: string | null
  old_progress: number | null
  new_progress: number | null
  note: string | null
  user: UserMin
  created_at: string
}

export interface DirectiveAttachmentRead {
  id: number
  filename: string
  file_size: number
  file_mime: string | null
  user: UserMin
  created_at: string
}

export interface DirectiveReadDetail extends DirectiveRead {
  units: DirectiveUnitRead[]
  linked_tasks: DirectiveTaskRead[]
  comments: DirectiveCommentRead[]
  history: DirectiveHistoryRead[]
  attachments: DirectiveAttachmentRead[]
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  size: number
  pages: number
}

export interface DirectiveCreate {
  title: string
  content?: string
  issuer_id: number
  status?: DirectiveStatus
  priority?: DirectivePriority
  issued_date?: string
  deadline?: string
  doc_id?: number | null
  assignee_staff_id?: number | null
  responsible_department_id?: number | null
}

export interface DirectiveUpdate extends Partial<DirectiveCreate> {}

export interface DirectiveUnitCreate {
  unit_name: string
  role?: string
  department_id?: number | null
  user_id?: number | null
  progress?: number
  note?: string
}

export interface DirectiveTaskCreate {
  title: string
  description?: string
  priority: string
  deadline?: string
  assignee_id?: number | null
  assignee_staff_id?: number | null
}

export interface DirectiveStats {
  total: number
  active: number
  completed: number
  overdue: number
  near_deadline: number
  avg_progress: number
}

File: backend/app/schemas/directive.py
from __future__ import annotations

from datetime import date, datetime
from math import ceil
from typing import Generic, TypeVar

from pydantic import AliasChoices, BaseModel, Field

T = TypeVar("T")


class PaginatedResponse(BaseModel, Generic[T]):
    items: list[T]
    total: int
    page: int
    size: int
    pages: int


class UserMin(BaseModel):
    id: int
    username: str
    full_name: str | None
    model_config = {"from_attributes": True}


class DocMin(BaseModel):
    id: int
    doc_number: str | None
    title: str
    doc_type: str
    model_config = {"from_attributes": True}


class StaffMin(BaseModel):
    id: int
    full_name: str
    position: str | None
    employee_code: str | None
    model_config = {"from_attributes": True}


class DeptMin(BaseModel):
    id: int
    name: str
    short_name: str | None
    model_config = {"from_attributes": True}


class TaskMin(BaseModel):
    id: int
    title: str
    status: str
    priority: str
    deadline: datetime | None = Field(None, validation_alias=AliasChoices("due_date", "deadline"))
    assignee: UserMin | None
    model_config = {"from_attributes": True, "populate_by_name": True}


# ─── Directive ────────────────────────────────────────────────────────────────

class DirectiveCreate(BaseModel):
    title: str
    content: str | None = None
    issuer_id: int
    status: str = "active"
    priority: str = "normal"
    issued_date: date | None = None
    deadline: datetime | None = None
    doc_id: int | None = None
    assignee_staff_id: int | None = None
    responsible_department_id: int | None = None
    coordinating_dept_ids: list[int] = []


class DirectiveUpdate(BaseModel):
    title: str | None = None
    content: str | None = None
    issuer_id: int | None = None
    priority: str | None = None
    issued_date: date | None = None
    deadline: datetime | None = None
    doc_id: int | None = None
    assignee_staff_id: int | None = None
    responsible_department_id: int | None = None
    coordinating_dept_ids: list[int] | None = None


class DirectiveStatusUpdate(BaseModel):
    status: str
    note: str | None = None


class DirectiveRead(BaseModel):
    id: int
    title: str
    content: str | None
    issuer_id: int
    status: str
    priority: str
    issued_date: date | None
    deadline: datetime | None
    progress: int
    doc_id: int | None
    assignee_staff_id: int | None
    responsible_department_id: int | None
    created_by: int
    issuer: UserMin
    creator: UserMin
    document: DocMin | None
    assignee_staff: StaffMin | None
    responsible_department: DeptMin | None
    created_at: datetime
    updated_at: datetime | None
    model_config = {"from_attributes": True}


# ─── Units ────────────────────────────────────────────────────────────────────

class DirectiveUnitCreate(BaseModel):
    unit_name: str
    role: str = "Thực hiện"
    department_id: int | None = None
    user_id: int | None = None
    progress: int = 0
    note: str | None = None


class DirectiveUnitUpdate(BaseModel):
    progress: int
    note: str | None = None


class DirectiveUnitRead(BaseModel):
    id: int
    unit_name: str
    role: str
    department_id: int | None
    user_id: int | None
    progress: int
    note: str | None
    department: DeptMin | None
    user: UserMin | None
    created_at: datetime
    updated_at: datetime | None
    model_config = {"from_attributes": True}


# ─── Tasks ────────────────────────────────────────────────────────────────────

class DirectiveTaskCreate(BaseModel):
    title: str
    description: str | None = None
    priority: str = "medium"
    deadline: datetime | None = None
    assignee_id: int | None = None
    assignee_staff_id: int | None = None


class DirectiveTaskRead(BaseModel):
    id: int
    task: TaskMin
    created_at: datetime
    model_config = {"from_attributes": True}


# ─── Comments ─────────────────────────────────────────────────────────────────

class DirectiveCommentCreate(BaseModel):
    content: str


class DirectiveCommentRead(BaseModel):
    id: int
    content: str
    user: UserMin
    created_at: datetime
    model_config = {"from_attributes": True}


# ─── History ──────────────────────────────────────────────────────────────────

class DirectiveHistoryRead(BaseModel):
    id: int
    action: str
    old_status: str | None
    new_status: str | None
    old_progress: int | None
    new_progress: int | None
    note: str | None
    user: UserMin
    created_at: datetime
    model_config = {"from_attributes": True}


# ─── Attachments ──────────────────────────────────────────────────────────────

class DirectiveAttachmentRead(BaseModel):
    id: int
    filename: str
    file_size: int
    file_mime: str | None
    user: UserMin
    created_at: datetime
    model_config = {"from_attributes": True}


# ─── Detail ───────────────────────────────────────────────────────────────────

class DirectiveReadDetail(DirectiveRead):
    units: list[DirectiveUnitRead] = []
    linked_tasks: list[DirectiveTaskRead] = []
    comments: list[DirectiveCommentRead] = []
    history: list[DirectiveHistoryRead] = []
    attachments: list[DirectiveAttachmentRead] = []


# ─── Stats ────────────────────────────────────────────────────────────────────

class DirectiveStats(BaseModel):
    total: int
    active: int
    completed: int
    overdue: int
    near_deadline: int
    avg_progress: float
